Relax ibgda synchronization

joalsop opened the pull request:
We can replace nearly all of the SEQ_CST atomic memory orderings with RELAXED in queue_pairs without impacting internode functionality. This results in 11% performance uplift for dispatch at the default input size (rdma BW increases from 43 GB/s - 48 GB/s) and greater relative improvements for smaller input sizes.

The only atomics that cannot be relaxed are:
1) A SEQ_CST system scope atomic in ring_doorbell (called within rocshmem_put), which ensures that previously stored data is flushed to memory (however, this can be relaxed without impacting functionality as long as the buffer is accessed uncached and there is a prior waitcnt).
2) An agent scope access by the leader thread in post_wqe_rma (also called within rocshmem_put). Relaxing this access causes internode to timeout. Further investigation can determine whether the waitcnt or the L1inv associated with this SEQ_CST ordering is necessary for functionality.

joalsop committed:
rocshmem mcm: relax ibdga orderings

convert all SEQ_CST orderings in queue_pair to RELAXED except:
-system scope ring_doorbell access: required to flush push buffer
 (unless data is uncached - in which case a waitcnt is sufficient)
-agent scope leader thread read in post_qpe_rma: unclear why this
 is necessary, but when relaxed, the code breaks. either the waitcnt
 or the L1inv associated with agent scope SEQ_CST is needed for
 functionality.

BKP commented on src/gpu_ib/queue_pair.cpp:
```diff
@@ -332,23 +332,23 @@ __device__ void QueuePair::quiet() {
       uint8_t mlx5_invld_bits = MLX5_CQE_INVALID << 4 | owner_bit;
       *((volatile uint8_t*)&cqe_entry->op_own) = mlx5_invld_bits;
-      __atomic_signal_fence(__ATOMIC_SEQ_CST);
+      __atomic_signal_fence(__ATOMIC_RELAXED);
```
SEQ_CST

BKP commented on src/gpu_ib/queue_pair.cpp:
```diff
@@ -332,23 +332,23 @@ __device__ void QueuePair::quiet() {
 
       swap_endian_store(const_cast<uint32_t*>(cq_dbrec), (uint32_t)(wave_cq_consumer + quiet_amount));
-      __atomic_signal_fence(__ATOMIC_SEQ_CST);
+      __atomic_signal_fence(__ATOMIC_RELAXED);
```
SEQ_CST

BKP commented on src/gpu_ib/queue_pair.cpp:
```diff
@@ -438,20 +438,20 @@ __device__ void QueuePair::post_wqe_rma(int pe, int32_t size, uintptr_t *laddr,
   seg_build.update_raddr_seg(raddr, rkey);
   seg_build.update_data_seg(laddr, size, lkey);
-  __atomic_signal_fence(__ATOMIC_SEQ_CST);
+  __atomic_signal_fence(__ATOMIC_RELAXED);
```
SEQ_CST

BKP commented on src/gpu_ib/queue_pair.cpp:
```diff
@@ -511,7 +511,7 @@ __device__ uint64_t QueuePair::post_wqe_amo(int pe, int32_t size, uintptr_t *rad
     quiet_internal(activemask, cons);
     ret = wave_fetch_atomic[my_logical_lane_id];
-    __atomic_signal_fence(__ATOMIC_SEQ_CST);
+    __atomic_signal_fence(__ATOMIC_RELAXED);
```
SEQ_CST

BKP commented on src/gpu_ib/queue_pair.cpp:
```diff
@@ -579,28 +579,28 @@ __device__ uint64_t QueuePair::post_wqe_amo(int pe, int32_t size, uintptr_t *rad
     seg_build.update_data_seg(nonfetching_atomic, 8, nonfetching_atomic_lkey);
   }
-  __atomic_signal_fence(__ATOMIC_SEQ_CST);
+  __atomic_signal_fence(__ATOMIC_RELAXED);
```
SEQ_CST

BKP commented on src/gpu_ib/queue_pair.cpp:
```diff
@@ -579,28 +579,28 @@ __device__ uint64_t QueuePair::post_wqe_amo(int pe, int32_t size, uintptr_t *rad
   if (fetching) {
     ret = wave_fetch_atomic[my_logical_lane_id];
-    __atomic_signal_fence(__ATOMIC_SEQ_CST);
+    __atomic_signal_fence(__ATOMIC_RELAXED);
```
SEQ_CST

BKP commented on src/gpu_ib/queue_pair.cpp:
```diff
@@ -255,12 +255,12 @@ __device__ void QueuePair::ring_doorbell(uint32_t pos) {
 __device__ void QueuePair::ring_doorbell(uint64_t db_val, uint64_t my_sq_counter) {
   swap_endian_store(const_cast<uint32_t*>(dbrec), (uint32_t)my_sq_counter);
-  __atomic_signal_fence(__ATOMIC_SEQ_CST);
+  __atomic_signal_fence(__ATOMIC_RELAXED);
```
Revert to SEQ_CST

BKP commented on src/gpu_ib/queue_pair.cpp:
```diff
@@ -215,7 +215,7 @@ __device__ void QueuePair::poll_wave_cqes(uint64_t activemask) {
   if (((cq_pos - cq_dbpos) & cq_mask) >= 100) {
     cq_dbpos = cq_pos;
-    __atomic_store_n(cq_dbreg, cq_dbval | (cq_mask & cq_dbpos), __ATOMIC_SEQ_CST);
+    __atomic_store_n(cq_dbreg, cq_dbval | (cq_mask & cq_dbpos), __ATOMIC_RELAXED);
```
Don't modify the code in the IONIC/THOR code path.

abouteiller commented:
The performance improvement from this PR is substantial, but there is no gain from the signal_fence change, I will push a commit that revert just that part 

Perf evaluation found here: 
https://amd.atlassian.net/wiki/spaces/MLSE/pages/1023161591/Performance+evaluation+of+relaxed+atomics+in+ibgda+conduit

abouteiller committed:
Undo changing atomic_signal_fence from SEQ_CST to RELAXED as this
appears to have no performance advantage and we are not entirely sure is
correct
